BepInEx.Core/Bootstrap/DefaultPluginProvider.cs

```diff
@@ -5,14 +5,25 @@
 using System.Reflection;
 using BepInEx.Logging;
 
-namespace BepInEx.PluginProvider;
+namespace BepInEx.Core.Bootstrap;
 
-[BepInPluginProvider("BepInExPluginProvider", "BepInExPluginProvider", "1.0")]
-internal class BepInExPluginProvider : BasePluginProvider
+internal class DefaultPluginProvider
 {
     private static readonly Dictionary<string, string> AssemblyLocationsByFilename = new();
     
-    public override IList<IPluginLoadContext> GetPlugins()
+    internal void Initialize()
+    {
+        Logger.Log(LogLevel.Message, "Started Initialise of default provider");
+        AppDomain.CurrentDomain.AssemblyResolve += (_, args) => ResolveAssembly(args.Name);
+        PhaseManager.Instance.OnPhaseStarted += phase =>
+        {
+            Logger.Log(LogLevel.Message, $"Providing on phase {phase}");
+            PluginManager.Instance.Providers.Add(new(), GetLoadContexts);
+        };
+        Logger.Log(LogLevel.Message, "Ended Initialise of default provider");
+    }
+
+    private IList<IPluginLoadContext> GetLoadContexts()
     {
         var loadContexts = new List<IPluginLoadContext>();
         foreach (var dll in Directory.GetFiles(Path.GetFullPath(Paths.PluginPath), "*.dll", SearchOption.AllDirectories))
```

Comment:
How about caching all of these contexts (as cached contexts) instead of reiterating over the files on every phase? Or do we expect dll files to be dropped at runtime? Wouldn't this add up across different providers having to reload on every phase?

Reply:
Added a discussion item in the OP because it depends if we expect the discovery to not change at entrypoint time